File: services/reservation-service/app/services/event_service.rb
class EventService
  include Singleton
  
  def initialize
    @connection = nil
    @channel = nil
    @exchange_name = ENV.fetch('RABBITMQ_EXCHANGE', 'library_events')
    connect
  end
  
  def self.publish_event(event_type, data)
    instance.publish_event(event_type, data)
  end
  
  def publish_event(event_type, data)
    return unless connected?
    
    event = {
      eventType: event_type,
      timestamp: Time.current.iso8601,
      source: 'reservation-service',
      data: data
    }
    
    routing_key = event_type.gsub('.', '_')
    
    @exchange.publish(
      event.to_json,
      routing_key: routing_key,
      persistent: true,
      content_type: 'application/json'
    )
    
    Rails.logger.info "Event published: #{event_type} with data: #{data}"
  rescue => e
    Rails.logger.error "Failed to publish event #{event_type}: #{e.message}"
  end
  
  def connected?
    @connection && @connection.open? && @channel && @channel.open?
  end
  
  def connect
    return if connected?
    
    begin
      rabbitmq_url = ENV.fetch('RABBITMQ_URL', 'amqp://localhost:5672')
      @connection = Bunny.new(rabbitmq_url)
      @connection.start
      
      @channel = @connection.create_channel
      @exchange = @channel.topic(@exchange_name, durable: true)
      
      Rails.logger.info "Connected to RabbitMQ at #{rabbitmq_url}"
    rescue => e
      Rails.logger.error "Failed to connect to RabbitMQ: #{e.message}"
      @connection = nil
      @channel = nil
      @exchange = nil
    end
  end
  
  def disconnect
    return unless connected?
    
    begin
      @channel&.close
      @connection&.close
      Rails.logger.info "Disconnected from RabbitMQ"
    rescue => e
      Rails.logger.error "Error disconnecting from RabbitMQ: #{e.message}"
    ensure
      @connection = nil
      @channel = nil
      @exchange = nil
    end
  end
  
  def reconnect
    disconnect
    connect
  end
  
  # Specific event publishing methods
  def self.publish_reservation_created(reservation)
    publish_event('reservation.created', {
      reservationId: reservation.id,
      userId: reservation.user_id,
      bookId: reservation.book_id,
      dueDate: reservation.due_date.iso8601,
      status: reservation.status
    })
  end
  
  def self.publish_reservation_returned(reservation)
    publish_event('reservation.returned', {
      reservationId: reservation.id,
      userId: reservation.user_id,
      bookId: reservation.book_id,
      returnDate: reservation.returned_at.iso8601,
      status: reservation.status
    })
  end
  
  def self.publish_reservation_overdue(reservation)
    publish_event('reservation.overdue', {
      reservationId: reservation.id,
      userId: reservation.user_id,
      bookId: reservation.book_id,
      dueDate: reservation.due_date.iso8601,
      daysOverdue: reservation.days_overdue,
      status: reservation.status
    })
  end
  
  def self.publish_reservation_extended(reservation, old_due_date)
    publish_event('reservation.extended', {
      reservationId: reservation.id,
      userId: reservation.user_id,
      bookId: reservation.book_id,
      oldDueDate: old_due_date.iso8601,
      newDueDate: reservation.due_date.iso8601,
      status: reservation.status
    })
  end
end

# Ensure proper cleanup on application shutdown
at_exit do
  EventService.instance.disconnect if defined?(EventService)
end 

File: services/user-service/src/services/eventService.js
const amqp = require("amqplib");

class EventService {
    constructor() {
        this.connection = null;
        this.channel = null;
        this.exchange = process.env.RABBITMQ_EXCHANGE || "library_events";
    }

    async connect() {
        try {
            const rabbitmqUrl =
                process.env.RABBITMQ_URL || "amqp://localhost:5672";
            this.connection = await amqp.connect(rabbitmqUrl);
            this.channel = await this.connection.createChannel();

            // Declare exchange
            await this.channel.assertExchange(this.exchange, "topic", {
                durable: true,
            });

            console.log("Connected to RabbitMQ");
        } catch (error) {
            console.error("Failed to connect to RabbitMQ:", error.message);
            // Don't throw error - service should work without message broker
        }
    }

    async publishEvent(eventType, data) {
        try {
            if (!this.channel) {
                console.warn(
                    "RabbitMQ not connected, skipping event publication"
                );
                return;
            }

            const event = {
                eventType,
                timestamp: new Date().toISOString(),
                source: "user-service",
                data,
            };

            const routingKey = eventType.replace(".", "_");
            await this.channel.publish(
                this.exchange,
                routingKey,
                Buffer.from(JSON.stringify(event)),
                { persistent: true }
            );

            console.log(`Event published: ${eventType}`, data);
        } catch (error) {
            console.error("Failed to publish event:", error.message);
        }
    }

    async publishUserRegistered(user) {
        await this.publishEvent("user.registered", {
            userId: user._id.toString(),
            email: user.email,
            firstName: user.firstName,
            type: "member",
        });
    }

    async publishUserProfileUpdated(user) {
        await this.publishEvent("user.profile_updated", {
            userId: user._id.toString(),
            email: user.email,
            firstName: user.firstName,
            lastName: user.lastName,
        });
    }

    async close() {
        try {
            if (this.channel) {
                await this.channel.close();
            }
            if (this.connection) {
                await this.connection.close();
            }
            console.log("RabbitMQ connection closed");
        } catch (error) {
            console.error("Error closing RabbitMQ connection:", error.message);
        }
    }
}

// Create singleton instance
const eventService = new EventService();

// Handle graceful shutdown
process.on("SIGINT", async () => {
    await eventService.close();
});

process.on("SIGTERM", async () => {
    await eventService.close();
});

module.exports = eventService;
